core/src/main/java/com/linecorp/armeria/internal/common/BodySubscriber.java

```diff
@@ -0,0 +1,301 @@
+        }
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstream != null) {
+            subscription.cancel();
+            return;
+        }
+        upstream = subscription;
+        if (cancelCalled) {
+            subscription.cancel();
+            return;
+        }
+        if (pendingRequests > 0) {
+            subscription.request(pendingRequests);
+        }
+    }
+
+    @Override
+    public void request(long n) {
+        if (n <= 0) {
+            upstreamMessage.abort(new IllegalArgumentException(
+                    "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+            return;
+        }
+        if (upstreamExecutor.inEventLoop()) {
+            request0(n);
+        } else {
+            upstreamExecutor.execute(() -> request0(n));
+        }
+    }
+
+    private void request0(long n) {
+        final Subscription upstream = this.upstream;
+        if (upstream == null) {
+            pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+        } else {
+            upstream.request(n);
+        }
+    }
+
+    @Override
+    public final void cancel() {
+        if (cancelCalled) {
+            return;
+        }
+        cancelCalled = true;
+        if (!notifyCancellation) {
+            downstream = NoopSubscriber.get();
+        }
+        completeTrailers(HttpHeaders.of());
+        final Subscription upstream = this.upstream;
+        if (upstream != null) {
+            upstream.cancel();
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject httpObject) {
+        if (httpObject instanceof HttpHeaders) {
+            final HttpHeaders trailers = (HttpHeaders) httpObject;
+            completeTrailers(trailers);
+            return;
+        }
+
+        assert httpObject instanceof HttpData;
+
+        final EventExecutor executor = downstreamExecutor;
+        if (executor.inEventLoop()) {
+            onNext0((HttpData) httpObject);
+        } else {
+            executor.execute(() -> onNext0((HttpData) httpObject));
+        }
+    }
+
+    private void onNext0(HttpData httpData) {
+        wroteAny = true;
+        if (!usePooledObject) {
+            httpData = PooledObjects.copyAndClose(httpData);
+        }
+        downstream.onNext(httpData);
+    }
+
+    @Override
+    public void onComplete() {
+        maybeCompleteHeaders(null);
+        final EventExecutor executor = downstreamExecutor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            completing = true;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onComplete0(downstream);
+        } else {
+            executor.execute(() -> onComplete0(downstream));
+        }
```

Comment:
As `upstreamExecutor` is specified as the `defaultSubscriberExecutor`, we can implement this code with some assumption that `downstreamExecutor` could be the same as `upstreamExecutor` with high probability.

```java
// in AbstactSplitHttpMessage
// volatile keyword might not be required.
private boolean needsDirectInvocation;

void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor, ...) {
    needsDirectInvocation = upstreamExecutor = executor;
}

boolean needsDirectInvocation() {
   if (needsDirectInvocation) {
      return true;
   }
   EventExecutor executor = this.executor;
   return executor != null && executor.inEventLoop();
}

// BodySubscriber
public void onComplete() {
    ...
    if (needsDirectInvocation()) { 
}
```

Reply:
Seems like a good idea! But I have a question.
Your opinion, is to add `needsDirectInvocation()` method(`protected`) to `AbstractSplitHttpMessage` and use it in `SplitHttpMessageSubscriber`(=`BodySubscriber`)? Or is it another way? I guess I didn't understand correctly.

Reply:
Ah, my comment was wrong. I wanted to make `needsDirectInvocation` non-volatile. How about invoking `bodySubscriber.initDownstream()` using `upstreamExecutor`?
```java
class AbstactSplitHttpMessage {
    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor, ...) {
        if (upstreamExecutor.inEventLoop()) {
            bodySubscriber.initDownstream(subscriber, executor, options);
        } else {
            upstreamExecutor.execute(() -> bodySubscriber.initDownstream(subscriber, executor, options));
        }
    }
}

class SplitHttpMessageSubscriber {
    private boolean needsDirectInvocation;

    void initDownstream(Subscriber<? super HttpData> downstream, EventExecutor downstreamExecutor, ...) {
        needsDirectInvocation = downstreamExecutor == upstreamExecutor;

        ....
        if (needsDirectInvocation) {
            downstream.onSubscribe(this);
            ...
        } else {
            downstreamExecutor.execute(() -> ...);
        }
    }

    public void onNext(HttpObject httpObject) {

        ...
        // needsDirectInvocation should be initialized before invoking onNext()
        if (needsDirectInvocation) {
            ...
        } else {

       
    }
}
```

Reply:
As you said, in the `subscribe` method of `AbstractSplitHttpMessage` class, `initDownstream` using `upstreamExecutor` seems to be correct.
In the https://github.com/line/armeria/pull/3953/commits/5db3add50a1f44d025beed56e6f18744558a4288 commit, some changes were made based on your comments.
Please feel free to let me if there is anything wrong. :)